### PR TITLE
Refactor bitesize10

### PR DESCRIPTION
**Pull request by t-young31**

### CODE REVIEW

File: tdms/matlabio/matlabio.h [hunk 1]
```diff
@@ -1,19 +1,78 @@
...
-unsigned char ***castMatlab3DArrayUint8(unsigned char *array, int nrows, int ncols, int nlayers);
-void freeCastMatlab3DArrayUint8(unsigned char ***castArray, int nlayers);
-int ***castMatlab3DArrayInt(int *array, int nrows, int ncols, int nlayers);
-void freeCastMatlab3DArrayInt(int ***castArray, int nlayers);
-int **castMatlab2DArrayInt(int *array, int nrows, int ncols);
-void freeCastMatlab2DArrayInt(int **castArray);
+/**
+ * Casts a 4-dimensional array such that it may be indexed according to the usual array indexing
+ * scheme array[l,k,j,i].
+ * @param array is a point to a matlab 4 dimensional array
```
Comment (samcunliffe) [inline tdms/matlabio/matlabio.h]:
```suggestion
 * @param array is a pointer to a matlab 4 dimensional array

```

File: tdms/matlabio/matlabio.h [hunk 2]
```diff
@@ -1,19 +1,78 @@
...
+ * @param nblocks the number of blocks, each of dimension nrows*ncols*nlayers
+ */
+double**** cast_matlab_4D_array(double *array, int nrows, int ncols, int nlayers, int nblocks);
+
+void free_cast_matlab_4D_array(double ****castArray, int nlayers, int nblocks);
+
+/**
+ * Casts a 3-dimensional array such that it may be indexed according to the usual array indexing
+ * scheme array[k,j,i].
+ * @param array is a point to a matlab 4 dimensional array
```
Comment (samcunliffe) [inline tdms/matlabio/matlabio.h]:
```suggestion
 * @param array is a pointer to a matlab 4 dimensional array

```

File: tdms/matlabio/matlabio.h [hunk 3]
```diff
@@ -1,19 +1,78 @@
...
+void free_cast_matlab_3D_array(T ***castArray, S nlayers){
+  for(S k=0; k<nlayers; k++)
+    free(castArray[k]);
+  free(castArray);
+}
+
+/**
+ * Casts a 2-dimensional array such that it may be indexed according to the usual array indexing
+ * scheme array[j,i].
+ * @param array is a point to a matlab 4 dimensional array
```
Comment (samcunliffe) [inline tdms/matlabio/matlabio.h]:
```suggestion
 * @param array is a pointer to a matlab 4 dimensional array

```

File: tdms/matlabio/matlabio.cpp [hunk 4]
```diff
@@ -238,16 +64,20 @@ void assert_is_struct_with_n_fields(const mxArray* ptr, int num, const std::stri
...
+  if (mxGetNumberOfDimensions(element) != n) {
+    throw runtime_error("Incorrect dimension on " + struct_name + "." + name + ". Required " +
+                        to_string(n) + "D");
   }
   return element;
 }
 
+mxArray* ptr_to_matrix_in(const mxArray* ptr, const string &name, const string &struct_name){
+  return ptr_to_nd_array_in(ptr, 2, name, struct_name);
+}
```
Comment (samcunliffe) [inline tdms/matlabio/matlabio.cpp:79]: This is a stylistic question... where a function simply calls another like this in one line. Do we do (or do you want to do) header-only?

File: tdms/matlabio/matlabio.h [hunk 5]
```diff
@@ -1,19 +1,78 @@
...
+ * scheme array[l,k,j,i].
+ * @param array is a point to a matlab 4 dimensional array
+ * @param nrows the number of rows in the array
+ * @param ncols the number of columns in the array
+ * @param nlayers the number of layers, each of dimension nrows*ncols
+ * @param nblocks the number of blocks, each of dimension nrows*ncols*nlayers
+ */
+double**** cast_matlab_4D_array(double *array, int nrows, int ncols, int nlayers, int nblocks);
+
+void free_cast_matlab_4D_array(double ****castArray, int nlayers, int nblocks);
```
Comment (samcunliffe) [inline tdms/matlabio/matlabio.h:22]:
```suggestion
/**
 * Frees the memory of a 4-dimensional array cast using cast_matlab_4D_array
 */
void free_cast_matlab_4D_array(double ****castArray, int nlayers, int nblocks);
```
I _think_ doxygen is smart enough to make this a link to the `cast_` function

File: tdms/matlabio/matlabio.h [hunk 6]
```diff
@@ -1,19 +1,78 @@
...
+  }
+  for(S k =0; k<nlayers; k++)
+    for(S j =0; j<ncols; j++){
+      p[k][j] = (array + k*nrows*ncols+ j*nrows);
+    }
+
+  return p;
+};
+
+template<typename T, typename S>
```
Comment (samcunliffe) [inline tdms/matlabio/matlabio.h:54]:
```suggestion
/**
 * Frees the memory of a 3-dimensional array cast using cast_matlab_3D_array
 */
template<typename T, typename S>
```

File: tdms/include/arrays.h [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+
+class FrequencyVectors{
+public:
+  Vector<double> x;
+  Vector<double> y;
+
+  void initialise(const mxArray *ptr);
+};
+
+class Pupil: public Matrix<double>{
```
Comment (samcunliffe) [inline tdms/include/arrays.h:160]: I'm a bit lost as to what this actually is.... really a →👁️  ?

**Review by samcunliffe — APPROVED** [review]
🔟 ✅

File: tdms/include/arrays.h (hunk 7 above)
Comment (t-young31) [reply tdms/include/arrays.h:160]: no idea 😞  – hence the lack of docstring – another one for Peter

Comment (t-young31) [inline tdms/include/arrays.h:160]:
```suggestion
// TODO: docstring
class Pupil: public Matrix<double>{
```

File: tdms/matlabio/matlabio.cpp (hunk 4 above)
Comment (t-young31) [reply tdms/matlabio/matlabio.cpp:79]: personal preference would be to have the definition in the header if either (1) it needs to be like some templated methods or (2) can fit on a single line i.e. `int size(){ return n; }` for instance

**Pull request merged**